timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/ExclusiveLock.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock;
+
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Queues;
+import com.palantir.atlasdb.timelock.util.LoggableIllegalStateException;
+import com.palantir.logsafe.SafeArg;
+
+import net.jcip.annotations.GuardedBy;
+
+public class ExclusiveLock implements AsyncLock {
+
+    @GuardedBy("this")
+    private final Queue<LockRequest> queue = Queues.newArrayDeque();
+    @GuardedBy("this")
+    private UUID currentHolder = null;
+
+    @Override
+    public synchronized CompletableFuture<Void> lock(UUID requestId) {
+        return submit(new LockRequest(requestId, false));
+    }
+
+    @Override
+    public CompletableFuture<Void> waitUntilAvailable(UUID requestId) {
+        return submit(new LockRequest(requestId, true));
+    }
+
+    private CompletableFuture<Void> submit(LockRequest request) {
+        queue.add(request);
+        processQueue();
+
+        return request.result;
+    }
+
+    @Override
+    public synchronized void unlock(UUID requestId) {
+        checkIsCurrentHolder(requestId);
+
+        currentHolder = null;
+        processQueue();
+    }
+
+    @VisibleForTesting
+    synchronized UUID getCurrentHolder() {
+        return currentHolder;
+    }
+
+    private void processQueue() {
+        while (!queue.isEmpty() && currentHolder == null) {
+            LockRequest head = queue.poll();
```

Comment:
Yeah, I went the route of making all public methods synchronized to start as I felt it was simpler / obviously correct. This is a pattern throughout the PR (although I broke it with `HeldLocks` in the last commit - I might revert that).